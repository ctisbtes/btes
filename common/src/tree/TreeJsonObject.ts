import { TreeNodeJsonObject } from './TreeNodeJsonObject';

export interface TreeJsonObject<TData> {
  root: TreeNodeJsonObject<TData> | null;
}
