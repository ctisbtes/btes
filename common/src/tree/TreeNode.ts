export class TreeNode<TData> {
  private _parent: TreeNode<TData> | null = null;
  public get parent(): TreeNode<TData> | null {
    return this._parent;
  }

  private _children: TreeNode<TData>[] = [];
  public get children(): ReadonlyArray<TreeNode<TData>> {
    return [...this._children];
  }

  public get hasParent(): boolean {
    return null !== this._parent;
  }

  public readonly id: string;
  public readonly data: TData;

  constructor(id: string, data: TData) {
    this.id = id;
    this.data = data;
  }

  /** 0-based distance from the root */
  public get height(): number {
    let count = 0;
    let currentNode = this.parent;

    while (currentNode !== null) {
      count++;
      currentNode = currentNode.parent;
    }

    return count;
  }

  public readonly setParent = (parent: TreeNode<TData>): void => {
    this._parent = parent;
  };

  public readonly addChild = (node: TreeNode<TData>): void => {
    this._children.push(node);
  };
}
