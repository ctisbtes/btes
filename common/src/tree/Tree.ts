/*
 * TODO:
 *
 * 1. If a `TreeNode` with children is added usin `add` method, internal state will get
 *    garbled (heads will not be properly updated). Do we even want an `add` method?
 *    Perhaps `create` is enough.
 *
 * 2. If we decide to keep the `add` method and support adding nodes with children,
 *    `importTreeNodeJsonObject` logic can be moved to the `TreeNode` class as a
 *    `TreeNode.fromJsonObject` static method. Then we can use that to construct the
 *    node chain, and import all of them at once via root node in `Tree.importTreeJsonObject`.
 *    That way we can get rid of the weird looking `importTreeNodeJsonObject` method as well.
 */

import _ from 'lodash';

import { TreeJsonObject } from './TreeJsonObject';
import { TreeNode } from './TreeNode';
import { TreeNodeJsonObject } from './TreeNodeJsonObject';

export class Tree<TData> {
  /**
   * Creates a new tree using the given `TreeJsonObject`.
   */
  public static readonly fromJsonObject = <TData>(
    treeJsonObject: TreeJsonObject<TData>
  ): Tree<TData> => {
    const tree = new Tree<TData>();
    tree.importTreeJsonObject(treeJsonObject, null);
    return tree;
  };

  /**
   * Imports the given `TreeJsonObject` into this tree at `mountPoint`.
   * @param nodeJsonObject The `TreeJsonObject` to be imported.
   * @param mountPoint Will be the parent of the imported tree. If `null`, root of the given `treeJsonObject` will be imported as the root of this tree.
   */
  public readonly importTreeJsonObject = (
    treeJsonObject: TreeJsonObject<TData>,
    mountPoint: TreeNode<TData> | null
  ): void => {
    if (null !== treeJsonObject.root) {
      this.importTreeNodeJsonObject(treeJsonObject.root, mountPoint);
    }
  };

  /**
   * Imports the given `TreeNodeJsonObject` with all its descendants into this tree at `mountPoint`.
   * @param nodeJsonObject The `TreeNodeJsonObject` to be imported along with all its descendants.
   * @param mountPoint Will be the parent of the imported tree. If `null`, given `nodeJsonObject` will be imported as the root of this tree.
   */
  public readonly importTreeNodeJsonObject = (
    nodeJsonObject: TreeNodeJsonObject<TData>,
    mountPoint: TreeNode<TData> | null
  ): void => {
    const current = this.createNode(
      nodeJsonObject.id,
      nodeJsonObject.data,
      mountPoint
    );

    for (const child of nodeJsonObject.children) {
      this.importTreeNodeJsonObject(child, current);
    }
  };

  /** Returns nodes from given array that are unique by id. */
  public static readonly uniq = <TData>(
    nodes: TreeNode<TData>[]
  ): TreeNode<TData>[] => {
    return _.uniqBy(nodes, (n) => n.id);
  };

  /** Returns whether given array has a node with given id. */
  public static readonly includesId = <TData>(
    nodes: readonly TreeNode<TData>[],
    nodeId: string
  ): boolean => {
    return nodes.some((n) => n.id === nodeId);
  };

  private _root: TreeNode<TData> | null = null;
  public get root(): TreeNode<TData> | null {
    return this._root;
  }

  private readonly _forkPoints: TreeNode<TData>[] = [];
  public get forkPoints(): ReadonlyArray<TreeNode<TData>> {
    return [...this._forkPoints];
  }

  private readonly _heads: TreeNode<TData>[] = [];
  public get heads(): ReadonlyArray<TreeNode<TData>> {
    return [...this._heads];
  }

  private _mainBranchHead: TreeNode<TData> | null = null;
  public get mainBranchHead(): TreeNode<TData> | null {
    return this._mainBranchHead;
  }

  public readonly createNode = (
    id: string,
    data: TData,
    parent: TreeNode<TData> | null
  ): TreeNode<TData> => {
    const node = new TreeNode<TData>(id, data);
    this.addNode(node, parent);
    return node;
  };

  public readonly addNode = (
    node: TreeNode<TData>,
    parent: TreeNode<TData> | null
  ): void => {
    if (node.hasParent) {
      throw new Error(
        'Trying to add a node that has a parent! Importing dirty nodes is not supported.' +
          'If you are trying to import JSON objects, use one of these: Tree.fromJsonObject`, `tree.importTreeJsonObject`, `tree.importTreeNodeJsonObject`.'
      );
    }

    if (node.children.length > 0) {
      throw new Error(
        'Trying to add a node that has children! Importing dirty nodes is not supported.' +
          'If you are trying to import JSON objects, use one of these: Tree.fromJsonObject`, `tree.importTreeJsonObject`, `tree.importTreeNodeJsonObject`.'
      );
    }

    if (parent === null) {
      if (this._root !== null) {
        throw new Error(
          'Trying to add a parentless node, but the tree already has a root!'
        );
      }

      this._root = node;
      this._heads.push(node);
      this._mainBranchHead = node;
    } else {
      if (parent.children.length > 0) {
        this.registerForkPoint(parent);
      }

      parent.addChild(node);
      node.setParent(parent);

      this.updateHeadsAfterAdd(parent.id, node);
    }
  };

  public readonly getNode = (nodeId: string): TreeNode<TData> | null => {
    return this.searchBackMultiple(nodeId, this._heads, []);
  };

  /** Returns the first fork point in this node's ancestors, or the root point if no fork points are found. */
  public readonly getForkPointOrRoot = (
    node: TreeNode<TData>
  ): TreeNode<TData> => {
    return this._getForkPointOrRootWithHeight(node).forkPointOrRoot;
  };

  public readonly isRootNode = (nodeId: string): boolean => {
    return this._root !== null && this._root.id === nodeId;
  };

  public readonly isForkPoint = (nodeId: string): boolean => {
    return Tree.includesId(this._forkPoints, nodeId);
  };

  public readonly isHeadNode = (nodeId: string): boolean => {
    return Tree.includesId(this._heads, nodeId);
  };

  /** 0 based distance of the given node from the first fork point or root. */
  public readonly getNodeHeightFromForkPointOrRoot = (
    node: TreeNode<TData>
  ): number => {
    return this._getForkPointOrRootWithHeight(node).height;
  };

  /** Returns an iterator that yields nodes. Starts from the head of main branch, goes backwards to the root. */
  public readonly getMainBranchIterator = (): Iterable<TreeNode<TData>> => {
    return this._mainBranchHead?.getIteratorToRoot() ?? [];
  };

  /** Returns an iterator that yields data. Starts from the head of main branch, goes backwards to the root. */
  public readonly getMainBranchDataIterator = (): Iterable<TData> => {
    return this._mainBranchHead?.getDataIteratorToRoot() ?? [];
  };

  /**
   * Returns a JSON serializable version of this tree with a nested object data structure.
   * `data` fields of all nodes will be directly included, so YOU need to make sure `TData`
   * is JSON seiralizable as well.
   */
  public readonly toJsonObject = (): TreeJsonObject<TData> => {
    return {
      root: this.root?.toJsonObject() || null,
    };
  };

  /**
   * Returns,
   * 1. The first fork point in this node's ancestors, or the root point if no fork points are found
   * 2. The 0-based height of the given node relative to the fork point or root
   */
  private readonly _getForkPointOrRootWithHeight = (
    node: TreeNode<TData>
  ): { forkPointOrRoot: TreeNode<TData>; height: number } => {
    let count = 0;
    let currentNode = node;

    // eslint-disable-next-line no-constant-condition
    while (
      !this.isRootNode(currentNode.id) &&
      !this.isForkPoint(currentNode.id)
    ) {
      if (null === currentNode.parent) {
        throw new Error('A node has no parent, but is not the root node!');
      }

      count++;
      currentNode = currentNode.parent;
    }

    return { height: count, forkPointOrRoot: currentNode };
  };

  /** Adds a node to fork points array if it is not already there. */
  private readonly registerForkPoint = (forkPoint: TreeNode<TData>): void => {
    if (!Tree.includesId(this._forkPoints, forkPoint.id)) {
      this._forkPoints.push(forkPoint);
    }
  };

  /**
   * Searches backwards from all `startNodes` until a node with `targetId` is found.
   * Does not search beyond fork points given in `ignoreForkPoints`.
   * @returns target node if found, `null` otherwise
   */
  private readonly searchBackMultiple = (
    targetId: string,
    startNodes: TreeNode<TData>[],
    ignoreForksPoints: TreeNode<TData>[]
  ): TreeNode<TData> | null => {
    const nextIgnoreForkPoints = ignoreForksPoints;
    const nextHeads: TreeNode<TData>[] = [];

    for (const startNode of startNodes) {
      const { stopType, stopNode } = this.searchBackUntilForkOrRoot(
        targetId,
        startNode
      );

      if (stopType === 'target') {
        return stopNode;
      } else if (stopType === 'fork') {
        if (null === stopNode.parent) {
          throw new Error('A node has no parent, but is not the root node!');
        }

        // do not visit the same fork point more than once
        if (!Tree.includesId(nextIgnoreForkPoints, stopNode.id)) {
          nextIgnoreForkPoints.push(stopNode);

          if (!Tree.includesId(nextHeads, stopNode.id)) {
            nextHeads.push(stopNode.parent);
          }
        }
      }
      // else if (stopType === 'root') {
      //   // there is nothing to search for beyond the root node, ignore
      // }
    }

    return nextHeads.length === 0
      ? null
      : this.searchBackMultiple(targetId, nextHeads, nextIgnoreForkPoints);
  };

  /** Searches backwards from `start` until it hits the target, a fork point, or the tree root. */
  private readonly searchBackUntilForkOrRoot = (
    targetId: string,
    startNode: TreeNode<TData>
  ): { stopType: 'root' | 'fork' | 'target'; stopNode: TreeNode<TData> } => {
    let currentNode = startNode;

    // eslint-disable-next-line no-constant-condition
    while (true) {
      if (currentNode.id === targetId) {
        return { stopType: 'target', stopNode: currentNode };
      }

      if (this.isRootNode(currentNode.id)) {
        return { stopType: 'root', stopNode: currentNode };
      }

      if (null === currentNode.parent) {
        throw new Error('A node has no parent, but is not the root node!');
      }

      if (this.isForkPoint(currentNode.id)) {
        return { stopType: 'fork', stopNode: currentNode };
      }

      currentNode = currentNode.parent;
    }
  };

  private readonly updateHeadsAfterAdd = (
    parentId: string,
    childNode: TreeNode<TData>
  ) => {
    _.remove(this._heads, (n) => n.id === parentId);
    this._heads.push(childNode);

    if (null === this._mainBranchHead) {
      throw new Error('Main branch head is null!');
    }

    // the use of `>` instead of `>=` is intentional
    // we don't want to switch if they have the same height
    if (childNode.height > this._mainBranchHead.height) {
      this._mainBranchHead = childNode;
    }
  };
}
