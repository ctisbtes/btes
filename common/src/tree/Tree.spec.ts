import { Tree } from './Tree';
import { TreeNode } from './TreeNode';
import { TreeJsonObject } from './TreeJsonObject';

interface NodeDataType {
  a: number;
  b: string;
}

const data: NodeDataType = {
  a: 1,
  b: 'foobar',
};

function makeComplexTree() {
  /*
   *                 c1 -> c2 -> c3
   *                /
   *         b1 -> b2 -> b3
   *        /
   * a1 -> a2 -> a3 -> a4 -> a5 -> a6
   *              \
   *               d1 -> d2 -> d3
   *                \
   *                 e1 -> e2 -> e3
   */

  const tree = new Tree<NodeDataType>();

  const a1 = tree.createNode('a1', data, null);
  const a2 = tree.createNode('a2', data, a1);
  const a3 = tree.createNode('a3', data, a2);
  const a4 = tree.createNode('a4', data, a3);
  const a5 = tree.createNode('a5', data, a4);
  const a6 = tree.createNode('a6', data, a5);
  const b1 = tree.createNode('b1', data, a2);
  const b2 = tree.createNode('b2', data, b1);
  const b3 = tree.createNode('b3', data, b2);
  const c1 = tree.createNode('c1', data, b2);
  const c2 = tree.createNode('c2', data, c1);
  const c3 = tree.createNode('c3', data, c2);
  const d1 = tree.createNode('d1', data, a3);
  const d2 = tree.createNode('d2', data, d1);
  const d3 = tree.createNode('d3', data, d2);
  const e1 = tree.createNode('e1', data, d1);
  const e2 = tree.createNode('e2', data, e1);
  const e3 = tree.createNode('e3', data, e2);

  return {
    tree,
    nodes: {
      a1,
      a2,
      a3,
      a4,
      a5,
      a6,
      b1,
      b2,
      b3,
      c1,
      c2,
      c3,
      d1,
      d2,
      d3,
      e1,
      e2,
      e3,
    },
  };
}

function idsOf<TData>(nodes: ReadonlyArray<TreeNode<TData>>) {
  return nodes.map((n) => n.id);
}

it('returns uniq nodes', () => {
  const nodeA = new TreeNode<NodeDataType>('nodeA', data);
  const nodeB = new TreeNode<NodeDataType>('nodeB', data);
  const nodeC = new TreeNode<NodeDataType>('nodeC', data);

  const r1 = Tree.uniq([nodeA, nodeB, nodeC]);
  expect(r1).toHaveLength(3);
  expect(r1).toContain(nodeA);
  expect(r1).toContain(nodeB);
  expect(r1).toContain(nodeC);

  const r2 = Tree.uniq([nodeA, nodeA, nodeA]);
  expect(r2).toHaveLength(1);
  expect(r2).toContain(nodeA);

  const r3 = Tree.uniq([]);
  expect(r3).toHaveLength(0);

  const r4 = Tree.uniq([nodeA, nodeA, nodeB, nodeB, nodeC, nodeC]);
  expect(r4).toHaveLength(3);
  expect(r4).toContain(nodeA);
  expect(r4).toContain(nodeB);
  expect(r4).toContain(nodeC);
});

it('determines if node array includes id', () => {
  const nodeA = new TreeNode<NodeDataType>('nodeA', data);
  const nodeB = new TreeNode<NodeDataType>('nodeB', data);
  const nodeC = new TreeNode<NodeDataType>('nodeC', data);
  const array = [nodeA, nodeB];

  expect(Tree.includesId(array, nodeA.id)).toBe(true);
  expect(Tree.includesId(array, nodeB.id)).toBe(true);
  expect(Tree.includesId(array, nodeC.id)).toBe(false);
});

it('has correct empty state', () => {
  const tree = new Tree<NodeDataType>();
  const outsiderNode = new TreeNode<NodeDataType>('node-id', data);

  expect(tree.forkPoints).toHaveLength(0);
  expect(tree.heads).toHaveLength(0);
  expect(tree.root).toBeNull();
  expect(tree.getNode(outsiderNode.id)).toBeNull();
  expect(() => tree.getForkPointOrRoot(outsiderNode)).toThrow();
  expect(tree.isHeadNode(outsiderNode.id)).toBe(false);
  expect(tree.isRootNode(outsiderNode.id)).toBe(false);
  expect(tree.isForkPoint(outsiderNode.id)).toBe(false);
});

it('creates root node', () => {
  const tree = new Tree<NodeDataType>();
  const rootNode = tree.createNode('root-node', data, null);

  expect(tree.forkPoints).toHaveLength(0);
  expect(tree.heads).toHaveLength(1);
  expect(tree.root).toBe(rootNode);
  expect(tree.getNode(rootNode.id)).toBe(rootNode);
  expect(tree.getForkPointOrRoot(rootNode)).toBe(rootNode);
  expect(tree.isHeadNode(rootNode.id)).toBe(true);
  expect(tree.isRootNode(rootNode.id)).toBe(true);
  expect(tree.isForkPoint(rootNode.id)).toBe(false);

  expect(rootNode.parent).toBeNull();
  expect(rootNode.hasParent).toBe(false);
  expect(rootNode.children).toHaveLength(0);
});

it('adds root node', () => {
  const tree = new Tree<NodeDataType>();
  const rootNode = new TreeNode<NodeDataType>('root-node', data);
  tree.addNode(rootNode, null);

  expect(tree.forkPoints).toHaveLength(0);
  expect(tree.heads).toHaveLength(1);
  expect(tree.root).toBe(rootNode);
  expect(tree.getNode(rootNode.id)).toBe(rootNode);
  expect(tree.getForkPointOrRoot(rootNode)).toBe(rootNode);
  expect(tree.isHeadNode(rootNode.id)).toBe(true);
  expect(tree.isRootNode(rootNode.id)).toBe(true);
  expect(tree.isForkPoint(rootNode.id)).toBe(false);

  expect(rootNode.parent).toBeNull();
  expect(rootNode.hasParent).toBe(false);
  expect(rootNode.children).toHaveLength(0);
});

it('gets node', () => {
  const tree = makeComplexTree().tree;

  const findAndAssertId = (nodeId: string) => {
    const foundNode = tree.getNode(nodeId);

    expect(foundNode).not.toBeNull();
    expect(foundNode?.id).toBe(nodeId);
  };

  [
    'a1',
    'a2',
    'a3',
    'a4',
    'a5',
    'a6',
    'b1',
    'b2',
    'b3',
    'c1',
    'c2',
    'c3',
    'd1',
    'd2',
    'd3',
    'e1',
    'e2',
    'e3',
  ].forEach(findAndAssertId);

  const notExistingNode = tree.getNode('foobar');
  expect(notExistingNode).toBeNull();
});

it('registers fork points', () => {
  /*
   *                 c1 -> c2 -> c3
   *                /
   *         b1 -> b2 -> b3
   *        /
   * a1 -> a2 -> a3 -> a4 -> a5 -> a6
   *              \
   *               d1 -> d2 -> d3
   *                \
   *                 e1 -> e2 -> e3
   */

  const tree = makeComplexTree().tree;
  const forkPointIds = tree.forkPoints.map((n) => n.id);

  expect(forkPointIds).toHaveLength(4);
  expect(forkPointIds).toContain('a2');
  expect(forkPointIds).toContain('b2');
  expect(forkPointIds).toContain('a3');
  expect(forkPointIds).toContain('d1');

  expect(tree.isForkPoint('a2')).toBe(true);
  expect(tree.isForkPoint('b2')).toBe(true);
  expect(tree.isForkPoint('a3')).toBe(true);
  expect(tree.isForkPoint('d1')).toBe(true);

  expect(tree.isForkPoint('a1')).toBe(false);
});

it('registers root', () => {
  const tree = makeComplexTree().tree;

  expect(tree.root).not.toBeNull();

  if (tree.root === null) {
    throw new Error('tree root should not be null here');
  }

  expect(tree.root.id).toBe('a1');
  expect(tree.isRootNode(tree.root.id)).toBe(true);
  expect(tree.isRootNode('a2')).toBe(false);
});

it('registers heads', () => {
  /*
   *                 c1 -> c2 -> c3
   *                /
   *         b1 -> b2 -> b3
   *        /
   * a1 -> a2 -> a3 -> a4 -> a5 -> a6
   *              \
   *               d1 -> d2 -> d3
   *                \
   *                 e1 -> e2 -> e3
   */

  const tree = makeComplexTree().tree;
  const headIds = tree.heads.map((n) => n.id);

  expect(headIds).toHaveLength(5);
  expect(headIds).toContain('c3');
  expect(headIds).toContain('b3');
  expect(headIds).toContain('a6');
  expect(headIds).toContain('d3');
  expect(headIds).toContain('e3');

  expect(tree.isHeadNode('c3')).toBe(true);
  expect(tree.isHeadNode('b3')).toBe(true);
  expect(tree.isHeadNode('a6')).toBe(true);
  expect(tree.isHeadNode('d3')).toBe(true);
  expect(tree.isHeadNode('e3')).toBe(true);

  expect(tree.isHeadNode('a1')).toBe(false);
});

it('finds root or fork points', () => {
  /*
   *                 c1 -> c2 -> c3
   *                /
   *         b1 -> b2 -> b3
   *        /
   * a1 -> a2 -> a3 -> a4 -> a5 -> a6
   *              \
   *               d1 -> d2 -> d3
   *                \
   *                 e1 -> e2 -> e3
   */

  const tree = makeComplexTree().tree;

  const getForkOrRoot = (id: string) => {
    const node = tree.getNode(id);
    if (!node) {
      throw Error('node should not be null!');
    }
    return tree.getForkPointOrRoot(node);
  };

  expect(getForkOrRoot('a1').id).toBe('a1');
  expect(getForkOrRoot('a2').id).toBe('a2');
  expect(getForkOrRoot('a3').id).toBe('a3');
  expect(getForkOrRoot('a4').id).toBe('a3');
  expect(getForkOrRoot('b1').id).toBe('a2');
  expect(getForkOrRoot('b2').id).toBe('b2');
  expect(getForkOrRoot('b3').id).toBe('b2');
  expect(getForkOrRoot('a5').id).toBe('a3');
  expect(getForkOrRoot('a6').id).toBe('a3');
  expect(getForkOrRoot('c1').id).toBe('b2');
  expect(getForkOrRoot('c2').id).toBe('b2');
  expect(getForkOrRoot('c3').id).toBe('b2');
  expect(getForkOrRoot('d1').id).toBe('d1');
  expect(getForkOrRoot('d2').id).toBe('d1');
  expect(getForkOrRoot('d3').id).toBe('d1');
  expect(getForkOrRoot('e1').id).toBe('d1');
  expect(getForkOrRoot('e2').id).toBe('d1');
  expect(getForkOrRoot('e3').id).toBe('d1');

  const outsiderNode = new TreeNode<NodeDataType>('outsider', data);
  expect(() => tree.getForkPointOrRoot(outsiderNode)).toThrow();
});

it('rejects parentless node if tree has root', () => {
  const tree = new Tree<NodeDataType>();
  tree.createNode('root-node', data, null);

  expect(() => tree.createNode('parentless', data, null)).toThrow();
});

it('calculates heights from fork points or root', () => {
  /*
   *                 c1 -> c2 -> c3
   *                /
   *         b1 -> b2 -> b3
   *        /
   * a1 -> a2 -> a3 -> a4 -> a5 -> a6
   *              \
   *               d1 -> d2 -> d3
   *                \
   *                 e1 -> e2 -> e3
   */
  const tree = makeComplexTree().tree;

  const getHeightFromForkOrRoot = (id: string) => {
    const node = tree.getNode(id);
    if (!node) {
      throw Error('node should not be null!');
    }
    return tree.getNodeHeightFromForkPointOrRoot(node);
  };

  expect(getHeightFromForkOrRoot('a1')).toBe(0);
  expect(getHeightFromForkOrRoot('a2')).toBe(0);
  expect(getHeightFromForkOrRoot('a3')).toBe(0);
  expect(getHeightFromForkOrRoot('a4')).toBe(1);
  expect(getHeightFromForkOrRoot('a5')).toBe(2);
  expect(getHeightFromForkOrRoot('a6')).toBe(3);
  expect(getHeightFromForkOrRoot('b1')).toBe(1);
  expect(getHeightFromForkOrRoot('b2')).toBe(0);
  expect(getHeightFromForkOrRoot('b3')).toBe(1);
  expect(getHeightFromForkOrRoot('c1')).toBe(1);
  expect(getHeightFromForkOrRoot('c2')).toBe(2);
  expect(getHeightFromForkOrRoot('c3')).toBe(3);
  expect(getHeightFromForkOrRoot('d1')).toBe(0);
  expect(getHeightFromForkOrRoot('d2')).toBe(1);
  expect(getHeightFromForkOrRoot('d3')).toBe(2);
  expect(getHeightFromForkOrRoot('e1')).toBe(1);
  expect(getHeightFromForkOrRoot('e2')).toBe(2);
  expect(getHeightFromForkOrRoot('e3')).toBe(3);

  const outsiderNode = new TreeNode<NodeDataType>('outsider', data);
  expect(() => tree.getNodeHeightFromForkPointOrRoot(outsiderNode)).toThrow();
});

it('converts to json object', () => {
  const tree = makeComplexTree().tree;
  expect(tree.toJsonObject()).toMatchSnapshot();
});

it('makes tree from json object', () => {
  const treeA = makeComplexTree().tree;
  const jsonObjA = treeA.toJsonObject();

  const treeB = Tree.fromJsonObject(jsonObjA);
  const jsonObjB = treeB.toJsonObject();

  expect(jsonObjA).toStrictEqual(jsonObjB);

  expect(idsOf(treeA.forkPoints)).toIncludeSameMembers(idsOf(treeB.forkPoints));
  expect(idsOf(treeA.heads)).toIncludeSameMembers(idsOf(treeB.heads));
  expect(treeA.root?.id).toBe(treeB.root?.id);
});

describe('guards against dirty node add', () => {
  it('throws on adding node with parent', () => {
    const { tree, nodes: treeNodes } = makeComplexTree();

    const parent = new TreeNode('parent', data);
    const child = new TreeNode('child', data);
    child.setParent(parent);
    parent.addChild(child);

    expect(() => tree.addNode(child, treeNodes.a6)).toThrow();
    expect(() => tree.addNode(child, treeNodes.a1)).toThrow();
    expect(() => tree.addNode(child, treeNodes.b2)).toThrow();
  });

  it('throws on adding node with children', () => {
    const { tree, nodes: treeNodes } = makeComplexTree();

    const parent = new TreeNode('parent', data);
    const child = new TreeNode('child', data);
    child.setParent(parent);
    parent.addChild(child);

    expect(() => tree.addNode(parent, treeNodes.a6)).toThrow();
    expect(() => tree.addNode(parent, treeNodes.a1)).toThrow();
    expect(() => tree.addNode(parent, treeNodes.b2)).toThrow();
  });

  it('throws on adding node with children and parent', () => {
    const { tree, nodes: treeNodes } = makeComplexTree();

    const parent = new TreeNode('parent', data);
    const node = new TreeNode('node', data);
    const child = new TreeNode('child', data);
    node.setParent(parent);
    child.setParent(node);
    parent.addChild(node);
    node.addChild(child);

    expect(() => tree.addNode(node, treeNodes.a6)).toThrow();
    expect(() => tree.addNode(node, treeNodes.a1)).toThrow();
    expect(() => tree.addNode(node, treeNodes.b2)).toThrow();
  });

  it('does not throw on adding clean node', () => {
    const { tree, nodes: treeNodes } = makeComplexTree();

    const node = new TreeNode('node', data);

    expect(() => tree.addNode(node, treeNodes.a6)).not.toThrow();
  });
});
