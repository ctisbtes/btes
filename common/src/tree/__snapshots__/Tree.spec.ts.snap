// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts to json object 1`] = `
Object {
  "root": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "data": Object {
                          "a": 1,
                          "b": "foobar",
                        },
                        "id": "a6",
                      },
                    ],
                    "data": Object {
                      "a": 1,
                      "b": "foobar",
                    },
                    "id": "a5",
                  },
                ],
                "data": Object {
                  "a": 1,
                  "b": "foobar",
                },
                "id": "a4",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "data": Object {
                          "a": 1,
                          "b": "foobar",
                        },
                        "id": "d3",
                      },
                    ],
                    "data": Object {
                      "a": 1,
                      "b": "foobar",
                    },
                    "id": "d2",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [],
                            "data": Object {
                              "a": 1,
                              "b": "foobar",
                            },
                            "id": "e3",
                          },
                        ],
                        "data": Object {
                          "a": 1,
                          "b": "foobar",
                        },
                        "id": "e2",
                      },
                    ],
                    "data": Object {
                      "a": 1,
                      "b": "foobar",
                    },
                    "id": "e1",
                  },
                ],
                "data": Object {
                  "a": 1,
                  "b": "foobar",
                },
                "id": "d1",
              },
            ],
            "data": Object {
              "a": 1,
              "b": "foobar",
            },
            "id": "a3",
          },
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [],
                    "data": Object {
                      "a": 1,
                      "b": "foobar",
                    },
                    "id": "b3",
                  },
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [
                          Object {
                            "children": Array [],
                            "data": Object {
                              "a": 1,
                              "b": "foobar",
                            },
                            "id": "c3",
                          },
                        ],
                        "data": Object {
                          "a": 1,
                          "b": "foobar",
                        },
                        "id": "c2",
                      },
                    ],
                    "data": Object {
                      "a": 1,
                      "b": "foobar",
                    },
                    "id": "c1",
                  },
                ],
                "data": Object {
                  "a": 1,
                  "b": "foobar",
                },
                "id": "b2",
              },
            ],
            "data": Object {
              "a": 1,
              "b": "foobar",
            },
            "id": "b1",
          },
        ],
        "data": Object {
          "a": 1,
          "b": "foobar",
        },
        "id": "a2",
      },
    ],
    "data": Object {
      "a": 1,
      "b": "foobar",
    },
    "id": "a1",
  },
}
`;
